core/src/main/java/org/jdbi/v3/core/Jdbi.java

```diff
@@ -444,15 +454,28 @@ public Handle open() {
      */
     public <R, E, X extends Exception> R withExtension(Class<E> extensionType, ExtensionCallback<R, E, X> callback)
             throws NoSuchExtensionException, X {
-        try (LazyHandleSupplier handle = new LazyHandleSupplier(this, config)) {
-            E extension = getConfig(Extensions.class)
-                    .findFor(extensionType, handle)
-                    .orElseThrow(() -> new NoSuchExtensionException("Extension not found: " + extensionType));
+        if (threadHandleSupplier.get() != null) {
+            return tryWithExtension(extensionType, callback, threadHandleSupplier.get());
+        }
 
-            return callback.withExtension(extension);
+        try (LazyHandleSupplier handleSupplier = new LazyHandleSupplier(this, config)) {
+            threadHandleSupplier.set(handleSupplier);
+            return tryWithExtension(extensionType, callback, handleSupplier);
+        } finally {
+            threadHandleSupplier.remove();
         }
     }
 
+    private <R, E, X extends Exception> R tryWithExtension(Class<E> extensionType,
```

Comment:
Maybe this is `callWithExtension`?  I don't see a `try` block or other failure handling here.